onnxruntime/core/providers/cuda/tensor/sequence_op.h

```diff
@@ -52,27 +51,37 @@ class SequenceConstruct final : public CudaKernel {
  public:
   SequenceConstruct(const OpKernelInfo& info) : CudaKernel(info) {}
   Status ComputeInternal(OpKernelContext* context) const override {
-    TensorSeq* Y = context->Output<TensorSeq>(0);
+    auto num_inputs = Node().InputArgCount().front();
+    ORT_ENFORCE(num_inputs >= 1, "Must have 1 or more inputs");
+
+    MLDataType first_dtype = context->Input<Tensor>(0)->DataType();
 
     AllocatorPtr alloc;
     ORT_ENFORCE(context->GetTempSpaceAllocator(&alloc).IsOK(),
                 "SequenceConstruct GPU: Unable to get an allocator.");
 
-    int32_t at = 0;
-    const Tensor* source_tensor = nullptr;
-    while (nullptr != (source_tensor = context->Input<Tensor>(at++))) {
-      if (1 == at) {
-        Y->SetType(source_tensor->DataType());
+    TensorSeq* Y = context->Output<TensorSeq>(0);
+    Y->SetType(first_dtype);
+    Y->Reserve(num_inputs);
+
+    for (int input_idx = 0; input_idx < num_inputs; ++input_idx) {
+      const auto* source_tensor = context->Input<Tensor>(input_idx);
+      if (input_idx > 0 && source_tensor->DataType() != first_dtype) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
```

Comment:
Do we need this check given that Add is already doing it? If we want to do this check we might as well do it for all num_inputs before creating tensors so that we don't incur the cost of memory allocation (due to Tensor::Create).

Reply:
Don't need it since Add() is already ensuring type consistency